### PR TITLE
Transform target dim

### LINKED ISSUE #614: [DOC] Depth to depth xgcm.Grid.transform onto terrain-following depth levels
**What were you looking for/trying to do?**:
Transform a field such as temp(x,y,z,t) from its original depths onto the non-uniform levels of a ROMS model grid.  This is a common step when preparing inputs for a nested model in which initial conditions and boundary conditions are sourced from a parent model. 

**What was missing from the docs or not clear enough?**:
The docs show how to do depth to depth transformation for 1-D spatially-uniform target depths but not for a 3-D array of depths.  From what I could tell, this does not seem to be within the standard functionality of `transform` and so requires a bit of workaround, but I could be wrong about this.

**How could the docs be improved? Additional examples, different wording, etc.**
This could either be a documentation improvement or a feature request (I chose the former because that's the easier solution).

I'd be willing to provide code for a minimal working example of how to do this within the current version of xgcm.  My method involves a 2-step transformation: first transforming source depths into destination sigma layer space, and then back into depths.

### PR DESCRIPTION
**Pull request by NoraLoose**
Still WIP

 - [ ] Closes #614
 - [ ] Documentation added
 - [ ] Tests added (WIP)
 - [ ] Passes `pre-commit run --all-files`
 - [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst` (WIP)

I added a section to the existing `doc/transform.ipynb` notebook to document the new capability. While working on this notebook, I also fixed pre-existing issues with this notebook including:

* Insert `autoparse_metadata=False` into `Grid()` to avoid error messages
* Correct dictionary keys for pangeo cloud data library
* Remove non-existing `xc` coordinate from Grid object to avoid existing `ValueError`
* Insert a `.squeeze()` for plotting AMOC to avoid error

### CODE REVIEW

**Comment by NoraLoose** [comment]
The changes made in this PR are not the reason for the failing checks above. The same checks fail in https://github.com/xgcm/xgcm/pull/641.

**Comment by NoraLoose** [comment]
Just realized that there is overlap between this PR and PR [#627](https://github.com/xgcm/xgcm/pull/627)

**Comment by jbusecke** [comment]
Hey @NoraLoose, thank you so much for working on this and sorry for the delay. I think I got the CI issues #644 and #648 under control now. Once #649 is merged can you try to rebase this one?

> Just realized that there is overlap between this PR and PR https://github.com/xgcm/xgcm/pull/627

Feel free to absorb and close this if needed.

**Comment by jbusecke** [comment]
Also I have restructured the linting and setup in #650, please let me know if it is ok to merge this ahead of time. I think this should not cause too big of a conflict?

**Comment by NoraLoose** [comment]
> Also I have restructured the linting and setup in https://github.com/xgcm/xgcm/pull/650, please let me know if it is ok to merge this ahead of time. I think this should not cause too big of a conflict?

Yep, feel free to merge https://github.com/xgcm/xgcm/pull/650!

**Comment by NoraLoose** [comment]
Transforming onto a spatially varying vertical coordinate (e.g., terrain-following) works with the linear interpolation method, see newly added tests and the example in the documentation.

I tested this also for the conservative interpolation method by adding the following test case to the `cases` dictionary in `xgcm/test/test_transform.py`:
```
    "conservative_depth_depth_multidim_target": {
        "source_coord": ("z", [5, 25, 60]),
        "source_bounds_coord": ("zc", [0, 10, 50, 75]),
        "source_data": ("data", [1, 4, 0]),
        "source_additional_data_coord": (
            "zc",
            [0, 10, 50, 75],
        ),
        "source_additional_data": ("zc", [0, 10, 50, 75]),
        # 2D target
        "target_dims": ("eta_rho", "s_w"),  # eta_rho: horizontal dimension; s_w: vertical dimension
        "target_coord": ("interface_depth_rho", np.array([[0, 1, 10, 50, 80], [0, 5, 20, 30, 100]])),
        "target_data": ("interface_depth_rho", np.array([[0, 1, 10, 50, 80], [0, 5, 20, 30, 100]])),
        "expected_dims": ("eta_rho", "s_w"),
        "expected_coord": ("interface_depth_rho", np.array([[0.5, 5.5, 30, 65], [2.5, 12.5, 25, 65]])),
        "expected_data": (
            "data",
            np.array([[0.1, 0.9, 4.0, 0.0], [0.5, 1.5, 1.0, 2.0]])
        ),
        "grid_kwargs": {
            "coords": {"Z": {"center": "z", "outer": "zc"}},
            "autoparse_metadata": False,
        },
        "transform_kwargs": {"method": "conservative", "target_data": "zc", "target_dim": "s_w"},
    }
```
However, it fails because of this line:
https://github.com/xgcm/xgcm/blob/2de04bfd36c376e782699e423dcdadb1053440f0/xgcm/transform.py#L160

**Comment by NoraLoose** [comment]
@jbusecke I'm not sure when I have time to look into generalizing the conservative method to multiple dimensions. We could merge this PR, and open a feature request issue? If we do this, should we include more warnings (where?) to clarify that the conservative method is not yet supported for multi-dimensional vertical target coordinates?

**Comment by jbusecke** [comment]
Thank you so much @NoraLoose! This is awesome. I am pretty busy today, but will try to take some time next week to review and maybe also enable conservative interpolation.

**Comment by NoraLoose** [comment]
@jbusecke I would like to use the new functionality of this PR in a downstream package ([roms-tools](https://github.com/CWorthy-ocean/roms-tools)). Do you think we can merge this soon, and leave the conservative method stuff for a future PR?

**Comment by NoraLoose** [comment]
@jbusecke @TomNicholas I was hoping to get this PR merged because a downstream package ([roms-tools](https://github.com/CWorthy-ocean/roms-tools)) is waiting on this new functionality. Any ideas who could review this PR if you guys are too busy?

**Comment by TomNicholas** [comment]
@NoraLoose I think at this point it's either @jbusecke or you just take over and use your own best judgement. This part of the code (the whole transform module) was @jbusecke 's work anyway. I will add you as a maintainer now regardless.

**Comment by NoraLoose** [comment]
@hdrake This PR is ready for review. It enables regridding onto a spatially varying vertical coordinate, such as a terrain-following coordinate.

I've added tests and included an example in the documentation, along with fixing some pre-existing issues in the `transform.ipynb` notebook.

The only remaining task is to enable conservative vertical regridding for the spatially varying coordinate, in addition to the linear method. I suggest opening an issue to track that feature, but merging this PR as is.

Let me know if you'd like to review or have any thoughts!

**Comment by hdrake** [comment]
@NoraLoose, I think fine to leave the conservative part for a later PR. I've added an explicit ValueError for the conservative case in the meantime.

I reviewed your updated notebook, which looks great! I also ran the tests (and skimmed the code, but found it a bit hard to follow at first glance) and can confirm everything passed locally.

Please feel free to merge when you're ready!

**Review by hdrake — APPROVED** [review]

**Comment by raphaeldussin** [comment]
LGTM +1
Very useful addition!

**Comment by NoraLoose** [comment]
Thanks for your input @hdrake and @raphaeldussin!

> skimmed the code, but found it a bit hard to follow at first glance

Yep, I remember it took me a while to understand the logic of the `test_transform.py` module as well.

I made three more modifications:
* Changed the `ValueError` that @hdrake added to a `NonImplementedError`
* Added a test that checks that the `NonImplementedError` is raised.
* Added some test data into the `test_transform.py` module that can be used to check the multi-dimensionsional conservative transform behaves correctly once that method is implemented.

**Comment by hdrake** [comment]
@NoraLoose, after adding `conda install pooch jupyterlab` to the `doc/environment.yml` environment, I was able to run your GLORYS -> ROMS example in the `transform.ipynb` notebook (but not the prior example because it requires access to the Google Cloud).

What environment are you using to run these example notebooks? I noticed, for example, that the notebook using `intake` which is not specified in either of `ci/environment.yml` or `doc/environment.yml`. Do the github actions that build the docs not actually run the notebooks? Do we need a seperate `doc/environment_notebooks.yml` to specify the environment that was used to create the example notebooks?

**Comment by NoraLoose** [comment]
@hdrake The GitHub Actions do not execute the notebooks; they are simply rendered as provided by the user.

I recall needing to add more packages to my environment to run the notebook from start to finish, but that environment isn’t clean. We should open a new issue to create a dedicated `doc/environment_notebooks.yml` with the necessary dependencies.

**Comment by NoraLoose** [comment]
I'm merging this, and will open two issues:
* Feature: Implement vertical **conservative** tranform for spatially-varying vertical coordinate
* Create dedicated `doc/environment_notebooks.yml`
